compiler/rustc_driver_impl/src/lib.rs

```diff
@@ -383,7 +383,31 @@ pub fn run_compiler(at_args: &[String], callbacks: &mut (dyn Callbacks + Send))
                 }
             }
 
-            Some(Linker::codegen_and_build_linker(tcx, &*compiler.codegen_backend))
+            // Hook for tests. This must be shortly before the
+            // `has_errors_or_delayed_bugs` check below for
+            // `tests/ui/treat-err-as-bug/span_delayed_bug.rs` to work.
+            if let Some((def_id, _)) = tcx.entry_fn(())
+                && tcx.has_attr(def_id, sym::rustc_delayed_bug_from_inside_query)
+            {
+                tcx.ensure_ok().trigger_delayed_bug(def_id);
+            }
+
+            // Don't emit metadata or do codegen if there were any errors.
+            //
+            // Note: checking for delayed bugs here is aggressive. If we have
+            // no errors but one or more delayed bugs, the `raise_fatal` call
+            // will instantly ICE when the DiagCtxt is dropped. I.e. this check
+            // serves as a kind of barrier, giving no opportunity for a
+            // subsequent error (e.g. during codegen) to nullify a delayed bug.
+            // The rationale is that delayed bugs are likely to cause more ICEs
+            // during codegen, obscuring the original problem.
+            if let Some(guar) = tcx.sess.dcx().has_errors_or_delayed_bugs() {
+                guar.raise_fatal();
+            }
```

Comment:
This way it is easy to forget to call this function before calling `codegen_and_build_linker`. rustc_interface is meant to be called directly by custom drivers if rustc_driver doesn't provide enough flexibility. For example for rustdoc. mutest-rs also uses it directly. Also by moving this call before `report_symbol_names`, errors in the latter no longer block codegen. Maybe instead just move the `tcx.sess.timings.start_section(tcx.sess.dcx(), TimingSection::Codegen);` later in `codegen_and_build_linker`?